scripts/topk.py

```diff
@@ -0,0 +1,639 @@
+    radix_position = dpt.asarray([n_bits_per_item - radix_bits], dtype=uint_type)
+    terminate = dpt.asarray([0], dtype=np.int32, device=sequential_device)
+    result = dpt.empty(sh=(k,), dtype=dtype, device=device)
+
+    if from_cpu_to_device:
+        check_radix_histogram = check_radix_histogram.configure(
+            sycl_queue=terminate.sycl_queue,
+            global_size=[1],
+            local_size=[1],
+        )
+
+    while True:
+        create_radix_histogram_kernel(
+            array_in_uint,
+            mask_for_desired_value,
+            desired_masked_value,
+            radix_position,
+            privatized_counts,
+        )
+
+        if privatized_counts.shape[0] == 1:
+            # TODO: improve sum to manage this case
+            counts = dpt.reshape(privatized_counts, (-1,))
+        else:
+            counts = dpt.reshape(reduce_privatized_counts(privatized_counts), (-1,))
+
+        if from_cpu_to_device:
+            counts = counts.to_device(sequential_device)
+            mask_for_desired_value = dpt.asarray(
+                mask_for_desired_value, device=sequential_device
+            )
+            desired_masked_value = dpt.asarray(
+                desired_masked_value, device=sequential_device
+            )
+            radix_position = dpt.asarray(radix_position, device=sequential_device)
+
+        check_radix_histogram(
+            counts,
+            k_in_subset,
+            radix_position,
+            mask_for_desired_value,
+            desired_masked_value,
+            terminate,
+        )
+
+        if from_cpu_to_device:
+            desired_masked_value = desired_masked_value.to_device(device)
+
+        if int(terminate[0]) == 1:
+            break
+
+        if from_cpu_to_device:
+            radix_position = radix_position.to_device(device)
+            mask_for_desired_value = mask_for_desired_value.to_device(device)
+            desired_masked_value = desired_masked_value.to_device(device)
+
+        initialize_privatized_counts(privatized_counts)
+
+    if from_cpu_to_device:
+        k_in_subset = k_in_subset.to_device(device)
+
+    # reinterpret the threshold back to a dtype item
+    threshold = dpt.usm_ndarray(
+        shape=desired_masked_value.shape, dtype=dtype, buffer=desired_masked_value
+    )
+    result_idx = dpt.zeros(sh=(1,), dtype=np.int32)
+
+    n_threshold_occurences = int(k_in_subset[0])
+
+    gather_topk_kernel = make_gather_topk_kernel(
+        n_items, k, n_threshold_occurences, work_group_size
+    )
+
+    gather_topk_kernel(array_in, k_in_subset, threshold, result_idx, result)
+    return result
+
+
+if __name__ == "__main__":
+    dtype = np.float32
+    size = 10
+    k = 3
+    array_in = dpt.asarray(np.random.default_rng().random(size), dtype=np.float32)
+    result = topk(array_in, k)
+    print(np.sort(dpt.asnumpy(result)))
+    print(np.sort(dpt.asnumpy(array_in))[-k:])
```

Comment:
does this comment still hold ? I favored hand crafted tests to go over all cases rather than random data. I don't think random data would add something new and it's less nice to debug when it fails :-/ (and for that the seed must be set so it's not so interesting to have random)